app/models/minion.rb

```diff
@@ -54,6 +63,18 @@ def self.assign_roles!(roles: {}, default_role: :minion)
     end
   end
 
+  def self.select_masters(roles)
```

Comment:
This method and the one below can be perfectly merged into a single method like:

```ruby
def self.select_role(ary, key, others: nil)
  if ary.key?(key)
    ary[key].map(&:to_i)
  else
    Minion.unassigned_role.pluck(:id) - others
  end
end
```

And I'm sure it can be further tuned if we merge things up in the main function 😉